src/basic/stat-util.c

```diff
@@ -266,3 +268,99 @@ int fd_verify_regular(int fd) {
 
         return stat_verify_regular(&st);
 }
+
+int device_path_make_major_minor(mode_t mode, dev_t devno, char **ret) {
+        const char *t;
+
+        /* Generates the /dev/{char|block}/MAJOR:MINOR path for a dev_t */
+
+        if (S_ISCHR(mode))
+                t = "char";
+        else if (S_ISBLK(mode))
+                t = "block";
+        else
+                return -ENODEV;
+
+        if (asprintf(ret, "/dev/%s/%u:%u", t, major(devno), minor(devno)) < 0)
+                return -ENOMEM;
```

Comment:
asprintf(3) says that `ret` is undefined on failure. So I think this should go through a level of inderection where `ret` is assigned only on success, and the pointer is freed if non-zero on failure internally here. This way the callers don't have to care about the implementation details on failure.

Reply:
we had this discussion in some github comments before. my reading of the man page is that it is undefined as in "not overwritten", i.e. set to whatever it was set before, whatever that might be. hence it's not really undefined, just as undefined as what was in there before.

This function is a GNU function and glibc does not override it on failure (why would it even?), hence I think we should be fine on this mostly, at least until they change their code on this.

Or to say this different: if we really have a problem here, then we need to fix things all over the place, and when we do that I think the better approach is to add asprintf_safe() or so, that gives our usual "no-clobber-on-failure" guarantee explicitly... and then patch our entire codebase for it.

But quite frankly, for now I think we don't really need to bother...

Reply:
I read "undefined" as meaning "either not set at all, or set to a valid pointer returned by malloc()". In other words, it's undefined whether the allocation occurred, but if anything is written to the pointer, it is a valid pointer. In the codebase, we always initialize the pointer to NULL and call free() on it, so we handle both possibilities correctly.